projects/ADuCM3029_demo_cn0428_cn0429/src/RingBuffer.c

```diff
@@ -0,0 +1,132 @@
+/**
+******************************************************************************
+*   @file     RingBuffer.c
+*   @brief    Ring buffer object source file
+*   @version  V0.1
+*   @author   ADI
+*
+*******************************************************************************
+*
+*Copyright 2015-2018(c) Analog Devices, Inc.
+*
+*All rights reserved.
+*
+*Redistribution and use in source and binary forms, with or without modification,
+*are permitted provided that the following conditions are met:
+*    - Redistributions of source code must retain the above copyright
+*      notice, this list of conditions and the following disclaimer.
+*    - Redistributions in binary form must reproduce the above copyright
+*      notice, this list of conditions and the following disclaimer in
+*      the documentation and/or other materials provided with the
+*      distribution.
+*    - Neither the name of Analog Devices, Inc. nor the names of its
+*      contributors may be used to endorse or promote products derived
+*      from this software without specific prior written permission.
+*    - The use of this software may or may not infringe the patent rights
+*      of one or more patent holders.  This license does not release you
+*      from the requirement that you obtain separate licenses from these
+*      patent holders to use this software.
+*    - Use of the software either in source or binary form, must be run
+*      on or directly connected to an Analog Devices Inc. component.
+*
+*THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+*INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS FOR A
+*PARTICULAR PURPOSE ARE DISCLAIMED.
+*
+*IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+*EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, INTELLECTUAL PROPERTY
+*RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+*BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+*STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+*THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************
+**/
+#include "Ringbuffer.h"
```

Comment:
I don't quite see the need to use the ringbuffer in this case. But it could just be me.

Reply:
I'm not sure what the original idea was. But conceptually, I think it's nice to have a structure where we can receive input while we're processing the previous command and it doesn't get lost like if you had a linear buffer where you clear everything and reset the pointer to the beginning. It's a simple enough wrapper for an array.

Reply:
Then this could be used as a direct buffer to store chars from the uart. It will make it as a central buffer and use the ring buffeers functions to parse commands. I was just suggesting removing the double parsing of strings once in cli and another time with the ring buffer functions.

Reply:
I understand the inefficiency in double parsing strings. But if it is at-all possible, I would like to avoid changing that. I would consider it a major change in the code structure that would change our schedule significantly and require a lot of testing. These boards are already available on the web and I don't want to keep customers waiting long for stylistic and optimization changes that don't affect functionality.